Makefile

```diff
@@ -204,11 +204,24 @@ $(if $(wildcard $(ITR)), \
 endif
 ELIB_DIR := $(realpath $(patsubst %/,%,$(patsubst %.,%,$(ELIB_DIR))))
 
+# IMPORT_EXCLSUFFIX (list of path suffixes to exclude)
```

Comment:
Hum... path suffix is then probably the wrong name. I am open for suggestions. What I wanted to say is that it is a suffix of a path that you are handing over for matching. On purpose it should match to the right part of a path so that it feels a bit like a relative path although it isn't. Because this suffix can be also a fully qualified path that matches from right to left, the idea was to get some convenience while being able to still do full path matching.

In general, I would have preferred a different method, like you say, match on the registered library name. However, we have currently the problem that a `Makefile.uk` is a simple file include. The library name is set by the libraries' `Makefile.uk` and you can in principle even register multiple libraries with the same file. In some rare cases this happens. If we are not following a very strict standard of `Makefile.uk` it can do whatever it wants and change any build system variable and rule. On one hand this gives us great flexibility but if you want to exclude a library from a build, I think the safest way is currently only by avoid including its `Makefile.uk`... and this works only based on the path.

I guess, longer term, we could solve it with per-library sub-make calls to create a per-library Make context, similar to the Linux kernel. Then libraries would have much more separated Make context and would not intefere with each other. Maybe this could solve restrictions that we have at the moment. But I think this is a more dramatic change and a bit of out of context of this PR.

Reply:
> Hum... path suffix is then probably the wrong name. I am open for suggestions. What I wanted to say is that it is a suffix of a path that you are handing over for matching. On purpose it should match to the right part of a path so that it feels a bit like a relative path although it isn't. Because this suffix can be also a fully qualified path that matches from right to left, the idea was to get some convenience while being able to still do full path matching.

Sorry I still don't understand what you mean by "suffix of a path". Later you state that "it should match the right part of a path", so in my understanding this means that it has to be either a relative path or an absolute path. But either way it is always a path. Perhaps you provide some examples to clarify the difference between a relative path, and absolute path and a "path suffix"?

Reply:
I am sorry that this is confusing, I am really open ears for a better naming. You need to consider it as a string matching operation. This means that even a substring of a folder name can be matched, as long it is the right part of the name.

Examples:

suffix of a path (fully qualified path): `$(UK_BASE)/lib/ukblkdev`
matches: `$(UK_BASE)/lib/ukblkdev`

suffix of a path (relative path): `lib/ukblkdev`
matches: `$(UK_BASE)/lib/ukblkdev`

suffix of a path (relative path): `ukblkdev`
matches: `$(UK_BASE)/lib/ukblkdev`

suffix of a path (substring): `blkdev`
matches: `$(UK_BASE)/lib/ukblkdev`
In detail:
```
path   $(UK_BASE)/lib/ukblkdev
suffix                  blkdev
```

suffix of a path (substring): `dev`
matches: `$(UK_BASE)/lib/uknetdev` `$(UK_BASE)/lib/ukblkdev`

suffix of a path (substring): `ukblk`
matches: none

Reply:
> You need to consider it as a string matching operation. This means that even a substring of a folder name can be matched, as long it is the right part of the name.

Ha! So `substring match` is the key here 🗝️  So effectively it's a substring match against the right-hand side of a path. Now it's clear, thanks.

My question is then, what is the motivation for a substring match? I am concerned of false positives, eg the user passes `dev` with the intention to match one library and then the system matches both `libukblkdev`, and `libuknetdev`, which makes no sense, as you need to have a one-to-one relationship with libraries that you want to drop and libraries you want to add. If it's there just for convenience (eg user wants to replace both `libukblkdev` and `libuknetdev`) I would find it more failsafe if we dropped substring match and require that the user explicitly specifies the libraries to exclude as paths, either absolute or relative to the unikraft's top-level directory, and internally convert these into fully qualified paths.

Reply:
The decision for this had to do with how implementable it is while fitting convenience cases. Matching a substring is the easiest method, a one liner operation in `Make` and `bash`. I guess, with some preparation we could get it to a mode where it at least matches a relative path, for example, we could prefix `/` whenever the path suffix does not start with `/`. However, as long as we do not resolve relative paths to fully qualified once, supporting relative paths with substring matching will always has corner cases and weird behavior: e.g., what is the expected base path (e.g., $(UK_ROOT), your current workdir), and what to do with `..` (imagine `lib/../lib/../lib/ukblkdev`).

A cleaner option could be to fix the input to a fully qualified path for now (same requirement as the other build system parameters as well). It would be less convenient to use by hand but then there is nothing unclear about what is filtered-out. If the path does not exist, it is also reasonable to throw a error message.

Reply:
I would say it's a reasonable requirement to express `E` relative to `$(UK_ROOT)`, since the libraries to exclude are always part of unikraft core.

In summary I would expect to see something in the application's Makefile along the lines of:
```
UK_ROOT ?= $(PWD)/../unikraft
UK_LIBS ?= $(PWD)/..

LIBS := $(UK_LIBS)/mynetdev
ELIBS := $(UK_ROOT)/lib/uknetdev

all:
	@$(MAKE) -C $(UK_ROOT) A=$(PWD) L=$(LIBS) E=$(ELIBS)

$(MAKECMDGOALS):
	@$(MAKE) -C $(UK_ROOT) A=$(PWD) L=$(LIBS) E=$(ELIBS) $(MAKECMDGOALS)
```

The reason I would avoid having the user to specify full paths is that you may build the same config in different systems, (ie locally and on the build server, and on a different CI/CD server for testing) each one of which may fetch sources under a different path, similarly to how all other paths in the application's `Makefile` are defined relative to `$(PWD)`.

Reply:
In your example you are calling Unikraft with fully qualified paths (or actually a path from root (`/`)) (see `ELIBS := $(UK_ROOT)/lib/uknetdev`). Except from what you write, I agree with the example because it is the approach also for the other parameters (`A`, `L`, ...). In the end there would no surprise if we make `E` inline with `L` etc. If you are okay, I'll make this change.

Reply:
Sounds great, thanks! In that case, I guess it's also clear how to adapt `config-submenu.sh`, update terminology etc.

Reply:
Yes, the behavior of both need to be in sync. 👍🏻